pkg/broker/ingress/stats_reporter.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ingress
+
+import (
+	"context"
+	"log"
+	"strconv"
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+	"knative.dev/pkg/metrics"
+	"knative.dev/pkg/metrics/metricskey"
+)
+
+// stats_exporter is adapted from knative.dev/eventing/pkg/broker/ingress/stats_reporter.go
+// with the following changes:
+// - Metric descriptions are updated to match GCP broker specifics.
+// - Pod name and container name tags are removed. These can cause high
+// cardinality with little value. Namespace and broker name tags are good enough
+// for identifying the metrics source.
+
+var (
+	// eventCountM is a counter which records the number of events received
+	// by the Broker.
+	eventCountM = stats.Int64(
+		"event_count",
+		"Number of events received by a Broker",
+		stats.UnitDimensionless,
+	)
+
+	// dispatchTimeInMsecM records the time spent dispatching an event to
+	// a Channel, in milliseconds.
+	dispatchTimeInMsecM = stats.Float64(
+		"event_dispatch_latencies",
+		"The time spent dispatching an event to the decouple topic",
+		stats.UnitMilliseconds,
+	)
+
+	// Create the tag keys that will be used to add tags to our measurements.
+	// Tag keys must conform to the restrictions described in
+	// go.opencensus.io/tag/validate.go. Currently those restrictions are:
+	// - length between 1 and 255 inclusive
+	// - characters are printable US-ASCII
+	namespaceKey         = tag.MustNewKey(metricskey.LabelNamespaceName)
+	brokerKey            = tag.MustNewKey(metricskey.LabelBrokerName)
+	eventTypeKey         = tag.MustNewKey(metricskey.LabelEventType)
+	responseCodeKey      = tag.MustNewKey(metricskey.LabelResponseCode)
+	responseCodeClassKey = tag.MustNewKey(metricskey.LabelResponseCodeClass)
+)
+
+type ReportArgs struct {
+	Namespace string
+	Broker    string
+	EventType string
+}
+
+func init() {
+	register()
+}
+
+// StatsReporter defines the interface for sending ingress metrics.
+type StatsReporter interface {
+	ReportEventCount(args *ReportArgs, responseCode int) error
+	ReportEventDispatchTime(args *ReportArgs, responseCode int, d time.Duration) error
+}
+
+var _ StatsReporter = (*reporter)(nil)
+var emptyContext = context.Background()
+
+// Reporter holds cached metric objects to report ingress metrics.
+type reporter struct{}
```

Comment:
I'm not sure we gain much by defining this reporter type vs. simply adding an event measure to handler and using the opencensus APIs directly w/ helper function if needed. This type being empty is also a clue the functionality could be expressed just as easily with helper functions or inlined. In my opinion this is an area where less abstraction would make the code easier to understand and modify.

Reply:
Sure I will revisit this. I simply copied the stats_reporter code from eventing without thinking too much about it.

Reply:
I don't feel strongly about this but I prefer having the `StatsReporter` interface and the `reporter` type which implements the interface. This allows us to switch to other implementations easily, e.g.,
* if we want to switch metrics to something else (OpenTelemetry in the future?)
* use a different implementation in tests (although I don't see a need for this yet).

Let me know your thoughts.

Reply:
If this was a general purpose interface then I would agree but since this interface is specific to this particular metric it doesn't actually make switching implementations any easier. So I think the relevant question is whether the interface that the stats reporter provides is more useful than the opencensus / otel APIs and I am not convinced that they are. The metrics APIs should also already be designed to be testable out of the box, see go.opencensus.io/metric/test.

Reply:
Changed this to a helper function, please take a look!